src/charm.py

```diff
@@ -0,0 +1,1017 @@
+            self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return True
+
+    @property
+    def _hydra_service_is_running(self) -> bool:
+        if not self._container.can_connect():
+            return False
+
+        try:
+            service = self._container.get_service(self._container_name)
+        except (ModelError, RuntimeError):
+            return False
+        return service.is_running()
+
+    @property
+    def _log_level(self) -> str:
+        return self.config["log_level"]
+
+    def _validate_config_log_level(self) -> bool:
+        is_valid = self._log_level in LOG_LEVELS
+        if not is_valid:
+            logger.info(f"Invalid configuration value for log_level: {self._log_level}")
+            self.unit.status = BlockedStatus("Invalid configuration value for log_level")
+        return is_valid
+
+    def _on_config_changed(self, event: ConfigChangedEvent) -> None:
+        """Event Handler for config changed event."""
+        self._handle_status_update_config(event)
+
+    @property
+    def _public_url(self) -> Optional[str]:
+        url = self.public_ingress.url
+        return normalise_url(url) if url else None
+
+    @property
+    def _admin_url(self) -> Optional[str]:
+        url = self.admin_ingress.url
+        return normalise_url(url) if url else None
+
+    @property
+    def _tracing_ready(self) -> bool:
+        return self.tracing.is_ready()
+
+    def _render_conf_file(self) -> str:
+        """Render the Hydra configuration file."""
+        with open("templates/hydra.yaml.j2", "r") as file:
+            template = Template(file.read())
+
+        secrets = self._get_secrets()
+        rendered = template.render(
+            cookie_secrets=[secrets["cookie"] if secrets else None],
+            system_secrets=[secrets["system"] if secrets else None],
+            log_level=self._log_level,
+            db_info=self._get_database_relation_info(),
+            consent_url=self._get_login_ui_endpoint_info("consent_url"),
+            error_url=self._get_login_ui_endpoint_info("oidc_error_url"),
+            login_url=self._get_login_ui_endpoint_info("login_url"),
+            hydra_public_url=self._public_url,
+            supported_scopes=SUPPORTED_SCOPES,
+        )
+        return rendered
+
+    def _set_version(self) -> None:
+        version = self._hydra_cli.get_version()
+        self.unit.set_workload_version(version)
+
+    def _get_database_relation_info(self) -> dict:
+        if not self.database.relations:
+            return None
+
+        relation_id = self.database.relations[0].id
+        relation_data = self.database.fetch_relation_data()[relation_id]
+        return {
+            "username": relation_data.get("username"),
+            "password": relation_data.get("password"),
+            "endpoints": relation_data.get("endpoints"),
+            "database_name": self._db_name,
+        }
```

Comment:
It would probably be better to check if the relation data contains the `username`, `password`, and `endpoints` fields, as the PostgreSQL charm may take some time to react and provision the database. During that time, the relation exists, but the relation data will be empty. Also, the `endpoints` field is a comma-separated list of all possible endpoints (e.g., `10.0.0.1:5432,10.0.0.2:5432`). Does Hydra support multiple host components within the URI?